RFC-0005-structured-kernel-definitions.md

```diff
@@ -0,0 +1,644 @@
+without requiring the rest of the system to be updated.
+
+One implication of this change is that the out parameter cannot be
+easily passed to existing public API that requires a mutable reference.
+This can be easily remedied by updating the existing APIs to accept
+const references and not only mutable references.
+
+### Type refinement
+
+Once we have dispatched to a CPU kernel, we know that the tensor in
+question is in fact a CPU tensor, and not (for example) a CUDA tensor.
+However, this information is not retained inside of the body of the
+kernel, and so if a user makes a method call or regular `at::` namespace
+function call, the dispatcher still must inspect the type tag to
+rediscover, yes, indeed, we still have a CPU tensor.
+
+One promising approach to solving this problem is to refine the type of
+a tensor from `Tensor` to `CPUTensor`, where a `CPUTensor` represents a
+tensor that is statically known to be a CPU Tensor.  Operations
+(functions and methods) on `CPUTensor` bypass dispatching and go
+directly to the CPU implementations in question.  `const CPUTensor&` can
+be defined to implicitly convert into `const Tensor&`, which means
+existing APIs that don't know how to short circuit can continue to do
+pre-existing behavior.
+
+The primary consequence of making this change immediately is we must
+immediately create a CPUTensor class with enough methods to cover the
+usual surface area (even if those methods don't apply any performance
+optimization).  With code generation this should not be too much code.
+This would also require the creation of a CPUTensorRef class to ensure
+that CPUTensors can be created from `const Tensor&` without incurring
+a reference count bump).
+
+One question is wheter or not the existence of CPUTensor means we should
+eliminate the `at::cpu::` namespace (as they serve near equivalent purposes;
+if you have functions which support CPUTensor, simply (unsafely) cast
+your Tensor to a CPUTensor and then utilize the regular API.)  One
+possible argument for retaining the `at::cpu::` namespace is that these
+functions are guaranteed to bypass dispatching, whereas other functions
+may implicitly downcast to `Tensor` and do an optimized call.
+
+## Long term status of unstructured kernels
+
+Structured operators are currently strictly defined in terms of an out
+operation.  However, there are some operators in PyTorch which do not
+have out variants, because they typically don't make sense.  Some of the
+most notable operator types of this form:
+
+* View operations
+* Factory functions
+* `copy_`
+
+Since non-structured kernels simply operate at a lower level of
+abstraction, in principle, it is not a big deal if some operators
+never become structured; to make an analogy, sometimes you have to
+write assembly, and as long as it is not too frequent, there is not
+too much to be gained from trying to extend the functionality of your
+system to expunge these entirely.
```

Comment:
A counterargument to consider here I think is that making structured kernels ubiquitous would allow other system capabilities to be defined in terms of them - e.g., shape analysis, memory planning.

The answer wouldn't be to torture everything into having in/func/out variants, but probably to add a handful of other little composite kernel system variations, each with its own codegen and triggering annotation in native_functions.yaml. Factory functions are an obvious fit for this kind of approach, but maybe they're the only good fit? Haven't thought it through any further.

Reply:
Well, this certainly *would* work.  What is less certain is if the number of operators that would be covered by the system variation warrant the variations. Well, at least we can delay designing this for later...